### ISSUE TITLE
High CPU consumption in the app

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/scope#1457]
On the service, with 4 nodes (probes) the app is over 100% CPU

<img width="1029" alt="screen shot 2016-05-10 at 07 34 46" src="https://cloud.githubusercontent.com/assets/2362916/15137421/c5955cbe-1681-11e6-8844-0efddeffda3e.png">

Profile: [pprof.localhost:4040.samples.cpu.001.pb.gz](https://github.com/weaveworks/scope/files/256507/pprof.localhost.4040.samples.cpu.001.pb.gz)

![app_cpu](https://cloud.githubusercontent.com/assets/2362916/15137570/9eeafb0e-1682-11e6-83dc-7abd85b1f9cd.png)

The profile shows 80% CPU (maybe I was seeing garbage collection spikes?)

Perhaps the cache removed on https://github.com/weaveworks/scope/pull/1447 would help?

**Comment by 2opremio** [comment]
This is not really a degradation compared to what we had before (see https://github.com/weaveworks/scope/issues/854 ) but not what I would had expected after #1418 )

**Comment by 2opremio** [comment]
Also, the pressure of gzip is pretty high (just like on #1454 ), we could try getting a better performant compressor library or reduce the compression level until we have a solution for #1201

**Comment by tomwilkie** [comment]
@2opremio what version was this?  Did it contain #1418?

**Comment by 2opremio** [comment]
Yep, as I mentioned both #1418 and #1447 are in

**Comment by 2opremio** [comment]
Reverting 2dae03501e741c6a5b841964ef13db3ed544c905 makes it even worse

<img width="1019" alt="screen shot 2016-05-10 at 09 32 15" src="https://cloud.githubusercontent.com/assets/2362916/15140266/3f2d38e8-1692-11e6-9f4c-fd5f4b73746d.png">

**Comment by 2opremio** [comment]
I've tried using https://github.com/klauspost/compress/gzip but it doesn't seem to be a big difference. @davkal , have the requests to `/api/topology` increased?

**Comment by janwillies** [comment]
We are seeing over 100% CPU usage in the app, with 5 probes connected.

Performance seems to be ok, though

This is with :latest from yesterday

**Comment by 2opremio** [comment]
> Performance seems to be ok, though

You mean that the app is usable?

**Comment by davkal** [comment]
> have the requests to /api/topology increased in this release?

Yes they have. Whenever you click on the search field, all topologies are fetched. But only then, no recurring timer.

**Comment by 2opremio** [comment]
@davkal I meant without search (this is master)

**Comment by davkal** [comment]
> I meant without search (this is master)

Not from the frontend. Just checked `git diff v0.14.0..origin/master -- client/app/scripts/utils/web-api-utils.js`. No refresh intervals changed.

**Comment by janwillies** [comment]
Yes the app is usable. We are using m3.large instances
Am 10.05.2016 12:48 schrieb "Alfonso Acosta" <notifications@github.com>:

**Comment by 2opremio** [comment]
@tomwilkie Could this be a bug on #1418 ?

**Comment by 2opremio** [comment]
I think I know what's happening. We started including container ENV variables which are huge in our service due to kubernetes services:

741 env variables per container 😮 which leads to well over 100K env variables in total:

<img width="1635" alt="screen shot 2016-05-10 at 13 11 53" src="https://cloud.githubusercontent.com/assets/2362916/15146233/211c7b50-16b2-11e6-9c41-ae7edd048cb2.png">

<img width="419" alt="screen shot 2016-05-10 at 13 13 21" src="https://cloud.githubusercontent.com/assets/2362916/15146243/2b83db2e-16b2-11e6-97d6-a6a29eeb7658.png">

I will try to truncate them and see if it helps.

**Comment by 2opremio** [comment]
@janwillies Can you check the number of env variables in your containers?

**Comment by janwillies** [comment]
170 per container, times ~100

2016-05-10 14:23 GMT+02:00 Alfonso Acosta <notifications@github.com>:

**Comment by 2opremio** [comment]
@janwillies Cool. Then truncating them might help.

**Comment by 2opremio** [comment]
@janwillies Reports that the app is consuming 129% CPU with 5 nodes

**Comment by willejs** [comment]
I am also seeing similar issues.

**Comment by 2opremio** [comment]
@willejs Can you give us some details about your setup? And, ideally, get as a CPU profile of the app?

You can easily do that by running

```
docker run --net=host -v $PWD:/root/pprof golang go tool pprof http://localhost:4040/debug/pprof/profile
```

In the host running the app

**Comment by idcrosby** [comment]
Attaching pprof data from our cluster on AWS seeing high CPU issue:
cc @2opremio 
[pprof.localhost:4040.samples.cpu.001.pb.gz](https://github.com/weaveworks/scope/files/318334/pprof.localhost.4040.samples.cpu.001.pb.gz)

**Comment by 2opremio** [comment]
This is the profile from @idcrosby, nothing too different from the previous profiles in this issue except for the futex contention on notesleep.

![app](https://cloud.githubusercontent.com/assets/2362916/16116813/e3855e10-33c6-11e6-8199-b55cefe793aa.png)

**Comment by 2opremio** [comment]
@janwillies @willejs @idcrosby Scope 0.17 is out and should improve the App's CPU consumption by at least 50% . Please let us know if the performance usage is acceptable now.

**Comment by alban** [comment]
> Also, the pressure of gzip is pretty high (just like on #1454 ), we could try getting a better performant compressor library or reduce the compression level until we have a solution for #1201

@2opremio I tried various compression level with https://github.com/kinvolk/scope/commit/37cc006fd58a0c58c4fbb6c74e12f2b93cb2cd2c and I get the following:

| Compression Level  | Report size in bytes | Time elapsed (*10) |
| ------------- | ------------- | ------------- |
| 0 (gzip.NoCompression) |  1568604 | 340ms |
| 1 (gzip.BestSpeed) | 114774 | 234ms |
| 2 | 97085 | 281ms |
| 3 | 94018 | 256ms |
| 4 | 91108 | 319ms |
| 5 | 85224 | 366ms |
| 6 (also using gzip.DefaultCompression == -1) | 80634 | 411ms |
| 7 | 79051 | 410ms |
| 8 | 78691 | 566ms |
| 9 (gzip.BestCompression) | 77781 | 618ms |

So if Scope were using `gzip.DefaultCompression` instead of `gzip.BestCompression`, the reports would be 4% bigger and compressed 33% faster.

I have not checked the decompression speed in the app.

**Comment by 2opremio** [comment]
Good job. I thought we were using the default compression. Also, it may be worth checking other compression libraries

**Comment by alban** [comment]
The list of compression https://golang.org/pkg/compress/

bzip2 is not suitable because it has only decompression and not compression (see https://github.com/golang/go/issues/4828)


File size: 1662592 (msgpack uncompressed)

| test name | level | compressed size | compression time | decompression time | 
| ------------- | -------------: | -------------: | -------------: | -------------: | 
| gzip | -1 | 143052 | 43761 | 8314 | 
| gzip | 0 | 1662745 | 1912 | 2494 | 
| gzip | 1 | 184074 | 15015 | 8360 | 
| gzip | 2 | 170205 | 15406 | 7611 | 
| gzip | 3 | 164116 | 22564 | 7013 | 
| gzip | 4 | 160214 | 33903 | 7382 | 
| gzip | 5 | 153660 | 35916 | 6992 | 
| gzip | 6 | 143052 | 46799 | 7414 | 
| gzip | 7 | 141689 | 56130 | 6422 | 
| gzip | 8 | 138420 | 107412 | 7927 | 
| gzip | 9 | 138127 | 170985 | 6594 | 
| zlib | -1 | 143040 | 45800 | 7895 | 
| zlib | 0 | 1662733 | 3070 | 2974 | 
| zlib | 1 | 184062 | 16524 | 11179 | 
| zlib | 2 | 170193 | 21051 | 9190 | 
| zlib | 3 | 164104 | 19900 | 8685 | 
| zlib | 4 | 160202 | 29980 | 9135 | 
| zlib | 5 | 153648 | 38100 | 9263 | 
| zlib | 6 | 143040 | 43449 | 8175 | 
| zlib | 7 | 141677 | 65277 | 7140 | 
| zlib | 8 | 138408 | 109087 | 7451 | 
| zlib | 9 | 138115 | 177843 | 7175 | 
| flate | -1 | 143034 | 48442 | 7317 | 
| flate | 0 | 1662727 | 1493 | 1379 | 
| flate | 1 | 184056 | 17335 | 8916 | 
| flate | 2 | 170187 | 15605 | 7419 | 
| flate | 3 | 164098 | 16493 | 7071 | 
| flate | 4 | 160196 | 30576 | 8306 | 
| flate | 5 | 153642 | 35537 | 7373 | 
| flate | 6 | 143034 | 43456 | 8109 | 
| flate | 7 | 141671 | 55900 | 6431 | 
| flate | 8 | 138402 | 106750 | 6652 | 
| flate | 9 | 138109 | 174859 | 6365 | 
| lzw | 0 | 551151 | 32472 | 23047 | 

I saved a msgpack file by using https://github.com/kinvolk/scope/commit/37cc006fd58a0c58c4fbb6c74e12f2b93cb2cd2c and I used https://gist.github.com/alban/4663d28f4cf69db4bb1d6dc365171e2c to produce this table.

Unfortunately, a lower compression level seems to imply a slower decompression. This is important since this bug is about the app.

In this table, flat with level=7 seems overall better for my msgpack file. But this is variable and when running the test another time, gzip with level=6 was the best overall.
